Add "copy to clipboard" button to code snippets

justinvp opened the pull request:
This PR adds a "Copy to Clipboard" button to code snippets.

![copybutton](https://user-images.githubusercontent.com/710598/61249508-43c4e700-a70a-11e9-81f8-4b436db2f6b8.gif)

Right now it adds it to any snippet that has a language specified, which covers the vast majority of uses throughout the site. There are cases that specify code snippets without a language, and right now those won't have a copy button. And I'm deliberately disabling it for the code snippets used for marketing content. We should prob follow-up and enable it across the board, and provide a first-class way to opt-out of the copy button when it isn't desired.

Copying `bash`, `shell`, and `sh`, snippets has some extra functionality. `$` prefixes are stripped, comments and empty lines are stripped, multiple lines are collapsed down to a single line, and any "output" lines that aren't prefixed with `$` are stripped.

Examples:

This:

```bash
$ mkdir hello-fargate && cd hello-fargate
$ pulumi new aws-typescript
```

...will have the this written to the clipboard:

```
mkdir hello-fargate && cd hello-fargate && pulumi new aws-typescript
```

And this:

```bash
$ curl http://$(pulumi stack output url)
<html>
    <head>
        <title>Hello Fargate</title>
    </head>
    <body>
        <p>Hello, containers!</p>
        <p>Made with ❤️ with <a href="https://pulumi.com">Pulumi</a></p>
    </body>
</html>
```

...will have this written to the clipboard:

```
curl http://$(pulumi stack output url)
```

Fixes #1023
Replaces #1204

justinvp commented on assets/sass/_fonts.scss:
```diff
@@ -1,2 +1,3 @@
 @import url('https://fonts.googleapis.com/css?family=Open+Sans:400,400i,400,700|Ubuntu:400,700');
 @import url('https://fonts.googleapis.com/css?family=PT+Mono');
+@import url('https://fonts.googleapis.com/css?family=Material+Icons');
```
I wanted to use the same material "copy" button we use on https://app.pulumi.com. As a follow-up, would anyone be opposed to removing our use of FontAwesome and just using the Material icons instead?

cnunciato replied:
Personally I’m open to having a look at how using Material icons affects the site overall, if that means being able to settle on a single icon set — but more generally, I’m honestly not in favor of propagating concrete Material design elements (as opposed to just guidance, which is fine) outside of the console.

justinvp commented on layouts/partials/code.html:
```diff
@@ -3,7 +3,7 @@
 {{ end }}
 
-<div class="code-mode-{{ default "light" .mode }} {{ if .chrome }} code-chrome {{ end }} code-{{ .lang }}">
+<div class="code-hero code-mode-{{ default "light" .mode }} {{ if .chrome }} code-chrome {{ end }} code-{{ .lang }}">
```
I added `code-hero` so I could use that in a selector to exclude the copy button from these. Though, I could see wanting to provide it for some uses. As a follow-up, we could add another option to enable it for this partial. (Would require some minor style updates for dark mode).

cnunciato replied:
Having `hero` in the classname bugs me a teeny bit (since there are lots of non-hero places where these are used), but it's only here, so definitely a nit. Maybe `.no-copy` (like `.no-select`), though? Up to you, just thinking aloud.

And yes, it'd be great to have these under Products as well! Could just surface them on hover, too.

justinvp commented on content/docs/reference/tutorials/kubernetes/tutorial-p8s-rollout.md:
````diff
@@ -73,54 +73,52 @@ can see `canary-example-app` and `canary-staging-app` created.
-
-   Permalink: https://app.pulumi.com/hausdorff/test-p8s/updates/1
-   ```
+        $ pulumi up
````
Workaround a bug in Hugo with code fences in lists. (As an aside, I think we can actually work around the bug by using `{{< highlight >}}` in the few places where the bug shows up, allowing us to upgrade to the latest version of Hugo. Will be looking at that separately.)

cnunciato commented on content/docs/reference/tutorials/kubernetes/tutorial-p8s-rollout.md:
````diff
@@ -55,16 +55,16 @@ can see `canary-example-app` and `canary-staging-app` created.
-   npm install
-   ```
+    ```sh
+    $ npm install
````
A few of the blocks on this page don't render the copy buttons, apparently because they're parented by an `li` and not a `div`. Changing the query to `:not(.code-hero) > div.highlight` seems to work, though, and _should_ be safe, I think.

justinvp commented:
@cnunciato, I addressed the feedback and made a few more tweaks. Mind taking another look?

Here's what it looks like using FontAwesome's `copy` icon (I got rid of the green checkmark after the click and went with just a tooltip to keep it simple):

![copy-fa1](https://user-images.githubusercontent.com/710598/61332512-42aebb00-a7d9-11e9-8430-b5d7be3e1148.gif)

![copy-fa-dark](https://user-images.githubusercontent.com/710598/61332520-45a9ab80-a7d9-11e9-94b6-4bd12d835533.gif)

Alternatively, we could use the `clipboard` icon, but I think I prefer `copy`.

<img width="468" alt="Screen Shot 2019-07-16 at 2 52 46 PM" src="https://user-images.githubusercontent.com/710598/61332564-68d45b00-a7d9-11e9-96f0-2f970618eb70.png">

cnunciato commented:
I like it! 👍
